benchmark/matrix_generator/matrix_generator.cpp

```diff
@@ -0,0 +1,199 @@
+#include <random>
+#include <vector>
+
+
+#include <gflags/gflags.h>
+#include <rapidjson/document.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/prettywriter.h>
+
+
+// some Ginkgo shortcuts
+using vector = gko::matrix::Dense<>;
+
+
+// helper for writing out rapidjson Values
+std::ostream &operator<<(std::ostream &os, const rapidjson::Value &value)
+{
+    rapidjson::OStreamWrapper jos(os);
+    rapidjson::PrettyWriter<rapidjson::OStreamWrapper> writer(jos);
+    value.Accept(writer);
+    return os;
+}
+
+
+// helper for splitting a comma-separated list into vector of strings
+std::vector<std::string> split(const std::string &s, char delimiter)
+{
+    std::istringstream iss(s);
+    std::vector<std::string> tokens;
+    for (std::string token; std::getline(iss, token, delimiter);
+         tokens.push_back(token))
+        ;
+    return tokens;
+}
+
+
+namespace {
+auto input_format =
+    "  [\n"
+    "    {\n"
+    "      \"filename\": \"<output-file>\",\n"
+    "      \"problem\": {\n"
+    "        \"type\": <matrix-type>\",\n"
+    "        \"num_blocks\": <num-blocks>,\n"
+    "        \"block_size\": <block-size>\n"
+    "      }\n"
+    "    }\n"
+    "  ]\n"
+    "  <output-file> is a string specifying a path to the output file\n"
+    "  <matrix-type> is a string specifying the type of matrix to generate,\n"
+    "    supported values are \"block-diagonal\"\n"
+    "  All other properties are optional, depending on <matrix-type>\n"
+    "  Properties for \"block-diagonal\":\n"
+    "    <num-blocks> is the number of dense diagonal blocks\n"
+    "    <block-size> is the size of each dense block\n"
+    "    The generated matrix will have a dense block of size <block-size>,\n"
+    "    with random real values chosen uniformly in the interval [-1, 1],\n"
+    "    repeated <num-blocks> times on the diagonal.\n";
+}  // namespace
+
+
+// input validation
+void print_config_error_and_exit(int code = 1)
+{
+    std::cerr << "Input has to be a JSON array of matrix configurations:\n"
+              << input_format << std::endl;
+    exit(code);
+}
+
+
+void validate_option_object(const rapidjson::Value &value)
+{
+    if (!value.IsObject() || !value.HasMember("filename") ||
+        !value["filename"].IsString() || !value.HasMember("problem") ||
+        !value["problem"].IsObject() || !value["problem"].HasMember("type") ||
+        !value["problem"]["type"].IsString()) {
+        print_config_error_and_exit(2);
+    }
+}
+
+
+// Command-line arguments
+DEFINE_uint32(seed, 42, "Seed used to generate the values of random matrices");
+
+
+void initialize_argument_parsing(int *argc, char **argv[])
+{
+    std::ostringstream doc;
+    doc << "A utility that generates various types of matrices.\n"
+        << "Usage: " << (*argv)[0] << " [options]\n"
+        << "  The standard input should contain a list of matrix\n"
+        << "  configurations in the following format:" << input_format
```

Comment:
minor: you may want to add a `\n` before `input_format` or directly in there so that this gets nicely formatted